app/models/minion.rb

```diff
@@ -9,37 +9,69 @@ class Minion < ApplicationRecord
   class NotEnoughMinions < StandardError; end
   # Raised when we fail to assign a role on a minion
   class CouldNotAssignRole < StandardError; end
+  # Raised when Minion doesn't exist
+  class NonExistingMinion < StandardError; end
 
   enum highstate: [:not_applied, :pending, :failed, :applied]
   enum role: [:master, :minion]
 
-  validates :hostname, uniqueness: true
+  validates :hostname, presence: true, uniqueness: true
 
-  # This method is used to assign the specified roles to Minions which do not
-  # have a roles already assigned (available minions).
-  # roles param are the roles we absolutely have to assign. If we can't assign
-  # one of those, the method will raise.
-  # default_role param can be set if we want all the rest of the available
-  # minions to get a default role.
-  # Returns the ids of the minions on which a roles has been assigned.
-  def self.assign_roles(roles: [], default_role: nil)
-    minions = Minion.where(role: nil)
-
-    # only the needed number of minions or all if we have a default role
-    minions = minions.limit(roles.size) unless default_role
-
-    raise NotEnoughMinions if minions.count < roles.size
+  # Example:
+  #   Minion.assign_roles(
+  #     roles: {
+  #       master: ["master.example.com"],
+  #       minion: ["minion1.example.com"]
+  #     },
+  #     default_role: :dns
+  #   )
+  # rubocop:disable Metrics/AbcSize
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
+  def self.assign_roles(roles: {}, default_role: :minion)
+    if roles[:master].any? && Minion.find_by(hostname: roles[:master].first).nil?
```

Comment:
Since we accept an `Array`, maybe we can do a best effort try and say something like `if roles[:master].any? && !Minion.exists?(hostname: roles[:master])`, so we ensure that we have at least one existing `master` assigned. With this logic we accept an `Array` but only take into account the first one into that given array.